File: laurus-ruby/test/test_search_batch.rb
# frozen_string_literal: true

require_relative "test_helper"

# Integration tests for Laurus::Index#search_batch
# (Phase 3d of #648, issue #719).
#
# Covers:
# - Empty input returns an empty Array without invoking the engine.
# - Single-query batch matches the single-query #search result.
# - Multi-query batch preserves input order and returns one result Array
#   per input query.
# - DSL strings, Query objects, and SearchRequest are all accepted.
class TestSearchBatch < Minitest::Test
  def create_index
    idx = Laurus::Index.new
    idx.put_document("doc1", { "title" => "Introduction to Rust", "body" => "Systems programming language." })
    idx.put_document("doc2", { "title" => "Python for Data Science", "body" => "Data analysis with Python." })
    idx.put_document("doc3", { "title" => "Distributed Systems", "body" => "Engineering at scale." })
    idx.commit
    idx
  end

  def test_search_batch_empty
    idx = create_index
    results = idx.search_batch([])
    assert_equal [], results
  end

  def test_search_batch_single_query_matches_search
    idx = create_index
    serial = idx.search("title:rust", limit: 5)
    batch = idx.search_batch(["title:rust"], limit: 5)

    assert_equal 1, batch.length
    assert_equal serial.length, batch[0].length
    batch[0].each_with_index do |result, i|
      assert_equal serial[i].id, result.id
    end
  end

  def test_search_batch_multi_query_preserves_order
    idx = create_index
    queries = ["title:rust", "body:python", "title:distributed"]
    expected_top_ids = %w[doc1 doc2 doc3]

    batch = idx.search_batch(queries, limit: 5)
    assert_equal queries.length, batch.length

    batch.each_with_index do |results, i|
      assert results.length >= 1, "expected at least 1 hit for query #{queries[i]}"
      assert_equal expected_top_ids[i], results[0].id
    end
  end

  def test_search_batch_with_query_objects
    idx = create_index
    queries = [
      Laurus::TermQuery.new("title", "rust"),
      Laurus::TermQuery.new("body", "python"),
    ]
    batch = idx.search_batch(queries, limit: 5)

    assert_equal 2, batch.length
    assert_equal "doc1", batch[0][0].id
    assert_equal "doc2", batch[1][0].id
  end

  def test_search_batch_with_search_requests
    idx = create_index
    requests = [
      Laurus::SearchRequest.new(lexical_query: Laurus::TermQuery.new("title", "rust"), limit: 5),
      Laurus::SearchRequest.new(lexical_query: Laurus::TermQuery.new("body", "python"), limit: 5),
    ]
    batch = idx.search_batch(requests)

    assert_equal 2, batch.length
    assert_equal "doc1", batch[0][0].id
    assert_equal "doc2", batch[1][0].id
  end

  def test_search_batch_no_match_returns_empty_inner_array
    idx = create_index
    queries = ["title:rust", "title:nonexistent_xyz"]
    batch = idx.search_batch(queries, limit: 5)

    assert_equal 2, batch.length
    assert batch[0].length >= 1
    assert_equal [], batch[1]
  end

  def test_search_batch_limit_per_query
    idx = create_index
    queries = ["body:programming OR body:data", "body:programming OR body:data"]
    batch = idx.search_batch(queries, limit: 1)

    assert_equal 2, batch.length
    batch.each do |results|
      assert results.length <= 1
    end
  end
end

File: laurus-python/tests/test_search_batch.py
"""Integration tests for ``Index.search_batch`` (Phase 3b of #648, issue #717).

These tests cover the Python-facing batched search API:

- Empty input returns an empty list without invoking the engine.
- Single-query batch matches the single-query ``search()`` result.
- Multi-query batch preserves input order and returns one result list
  per input query.
- DSL strings, ``Query`` objects, and ``SearchRequest`` are all accepted.
"""

import pytest
import laurus


@pytest.fixture
def index():
    """Return a fresh in-memory index with three indexed documents."""
    idx = laurus.Index()
    idx.put_document("doc1", {"title": "Introduction to Rust", "body": "Systems programming language."})
    idx.put_document("doc2", {"title": "Python for Data Science", "body": "Data analysis with Python."})
    idx.put_document("doc3", {"title": "Distributed Systems", "body": "Engineering at scale."})
    idx.commit()
    return idx


def test_search_batch_empty(index):
    """Empty list returns empty list."""
    results = index.search_batch([])
    assert results == []


def test_search_batch_single_query_matches_search(index):
    """search_batch([q]) should match search(q) for an equivalent query."""
    serial = index.search("title:rust", limit=5)
    batch = index.search_batch(["title:rust"], limit=5)
    assert len(batch) == 1
    assert len(batch[0]) == len(serial)
    for b, s in zip(batch[0], serial):
        assert b.id == s.id


def test_search_batch_multi_query_preserves_order(index):
    """Multi-query batch returns one list per input, in input order."""
    queries = ["title:rust", "body:python", "title:distributed"]
    expected_top_ids = ["doc1", "doc2", "doc3"]

    batch = index.search_batch(queries, limit=5)
    assert len(batch) == len(queries)

    for results, expected_id in zip(batch, expected_top_ids):
        assert len(results) >= 1, f"expected at least 1 hit for query targeting {expected_id}"
        assert results[0].id == expected_id


def test_search_batch_with_query_objects(index):
    """search_batch accepts ``Query`` objects, not just DSL strings."""
    queries = [
        laurus.TermQuery("title", "rust"),
        laurus.TermQuery("body", "python"),
    ]
    batch = index.search_batch(queries, limit=5)
    assert len(batch) == 2
    assert batch[0][0].id == "doc1"
    assert batch[1][0].id == "doc2"


def test_search_batch_with_search_requests(index):
    """search_batch accepts pre-built ``SearchRequest`` instances."""
    requests = [
        laurus.SearchRequest(lexical_query=laurus.TermQuery("title", "rust"), limit=5),
        laurus.SearchRequest(lexical_query=laurus.TermQuery("body", "python"), limit=5),
    ]
    batch = index.search_batch(requests)
    assert len(batch) == 2
    assert batch[0][0].id == "doc1"
    assert batch[1][0].id == "doc2"


def test_search_batch_no_match_returns_empty_inner_list(index):
    """A query with no matches yields an empty inner list, not a missing entry."""
    queries = ["title:rust", "title:nonexistent_xyz"]
    batch = index.search_batch(queries, limit=5)
    assert len(batch) == 2
    assert len(batch[0]) >= 1
    assert batch[1] == []


def test_search_batch_limit_per_query(index):
    """``limit`` applies to each query independently."""
    queries = ["body:programming OR body:data", "body:programming OR body:data"]
    batch = index.search_batch(queries, limit=1)
    assert len(batch) == 2
    for results in batch:
        assert len(results) <= 1
